clang-tools-extra/clang-tidy/modernize/UseIntegerSignComparisonCheck.h

```diff
@@ -0,0 +1,47 @@
+//===--- UseIntegerSignComparisonCheck.h - clang-tidy -----------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CLANG_TOOLS_EXTRA_CLANG_TIDY_MODERNIZE_USEINTEGERSIGNCOMPARISONCHECK_H
+#define LLVM_CLANG_TOOLS_EXTRA_CLANG_TIDY_MODERNIZE_USEINTEGERSIGNCOMPARISONCHECK_H
+
+#include "../ClangTidyCheck.h"
+#include "../utils/IncludeInserter.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+
+namespace clang::tidy::modernize {
+
+/// Class detects comparisons between signed and unsigned integers
+///
+/// For the user-facing documentation see:
+/// http://clang.llvm.org/extra/clang-tidy/checks/modernize/use-integer-sign-comparison.html
+class UseIntegerSignComparisonCheck : public ClangTidyCheck {
+public:
+  UseIntegerSignComparisonCheck(StringRef Name, ClangTidyContext *Context);
+
+  void storeOptions(ClangTidyOptions::OptionMap &Opts) override;
+  void registerPPCallbacks(const SourceManager &SM, Preprocessor *PP,
+                           Preprocessor *ModuleExpanderPP) override;
+  void registerMatchers(ast_matchers::MatchFinder *Finder) override;
+  void check(const ast_matchers::MatchFinder::MatchResult &Result) override;
+  bool isLanguageVersionSupported(const LangOptions &LangOpts) const override {
+    return LangOpts.CPlusPlus;
+  }
```

Comment:
Actually, there is a following check in the FixIt:
if (!(getLangOpts().CPlusPlus17 && IsQtApplication) &&
    !getLangOpts().CPlusPlus20)
return;
My idea was, if C++ version is less C++17 (like C++11), the FixIt still will highlight the comparison problem, but will not provide a replacement. So user will be able to fix a problem place manually with own means.
Does it make sense?   
(If not, I will change according the comment)

Reply:
Yep, my bad

Reply:
> the FixIt still will highlight the comparison problem, but will not provide a replacement.

I'm not sure I agree with this. Checks must be actionable. If a user does not have language support for fixing this, the check should not warn.

It's common practice for `modernize` checks to only be active in the C++ Standard they apply to. 

I see that the check no longer has any Qt dependencies, so we should simply ensure this check only runs in C++20 mode

Reply:
But based on that information the user can fix the problem manually in any c++ version. 
For example, the user can change type of variables, or add a check that 'int' value is not negative in his comparison statements. 
Is not it useful?

Reply:
The purpose of `modernize` checks is typically to transform the codebase into a new feature that is more modern (or it's "best practice") and makes the code simpler (that's why all except one have automatic fixes).

If we don't provide a fix, users will do "whatever they can" in pre C++20 to get rid of the warning. The quickest fix is to litter the code with `static_casts`, which is pretty bad because then the problem is silenced and they can't truly modernize their codebase automatically when they do move to C++20 (let alone introduce bugs).

Also, I expect this check to be very noisy in a real-world codebase, not quite fixable manually. 

Besides, "warning on integer comparison" already exists as a warning in Clang (`-Wsign-compare`), so this check would be redundant pre-C++20.

Reply:
If clang provide the warning, then it should not be checked in clang-tidy.

`if (!(getLangOpts().CPlusPlus17 && IsQtApplication) &&
!getLangOpts().CPlusPlus20)` for this case, I am not sure whether we should consider the common used c++ huge library like QT / boost. I think in the first version, maybe limit it to `CPlusPlus20` will be better.